src/features/wasm-bigint.md

```diff
@@ -0,0 +1,136 @@
+
+We can build that with
+
+```
+emcc example.c --js-library example.js -o out.js
+```
+
+When we run it, we don’t get what we expect:
+
+```
+node out.js
+JS received: 0x12345678
+```
+
+We sent `0xABCD12345678` but we only received `0x12345678` 😔. What happens here is that legalization turns that `i64` into two `i32`s, and our code just received the low 32 bits, and ignored another parameter that was sent. To handle things properly, we’d need to do something like this:
+
+```javascript
+  // The i64 is split into two 32-bit parameters, “low” and “high”.
+  send_i64_to_js: function(low, high) {
+    console.log("JS received: 0x" + high.toString(16) + low.toString(16));
+  }
+```
+
+Running this now, we get
+
+```
+JS received: 0xabcd12345678
+```
+
+As you can see, it’s possible to live with legalization. But it can be kind of annoying!
+
+## The solution: JavaScript BigInts
+
+JavaScript has [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) values now, which represent integers of arbitrary size, so they can represent 64-bit integers properly. It is natural to want to use those to represent `i64`s from Wasm. That’s exactly what the JS-BigInt-Integration feature does!
+
+Emscripten has support for Wasm BigInt integration, which we can use to compile the original example (without any hacks for legalization), by just adding `-s WASM_BIGINT`:
+
+```
+emcc example.c --js-library example.js -o out.js -s WASM_BIGINT
+```
+
+We can then run it (note that we need to pass Node.js a flag to enable BigInt integration currently):
+
+```
+node --experimental-wasm-bigint a.out.js
+JS received: 0xabcd12345678
+```
+
+Perfect, exactly what we wanted!
+
+And not only is this simpler, but it’s faster. As mentioned earlier, in practice it’s rare that `i64` conversions happen on a hot path, but when it does the slowdown can be noticeable. If we turn the above example into a benchmark, running many calls of `send_i64_to_js`, then the BigInt version is 18% faster.
+
+Another benefit is that when using BigInt integration the toolchain can avoid performing legalization on the Wasm, which makes building faster. In Emscripten, legalization happens after the LLVM linker (`wasm-ld`) runs, and so if we can avoid legalization we may be able to avoid making any changes to the Wasm file at all. You can get that speedup if you build with `-s WASM_BIGINT`, and do not provide any other flags that require changes to be made. For example, `-O0 -s WASM_BIGINT`, or `-O1 -s WASM_BIGINT`, will work (but at higher optimization levels [the Binaryen optimizer will run](https://emscripten.org/docs/optimizing/Optimizing-Code.html#link-times)).
```

Comment:
I might be wrong (not a native speaker either), but

```suggestion
Another benefit is that, when using the BigInt integration, the toolchain can avoid performing legalization on the Wasm, which makes building faster. In Emscripten, legalization happens after the LLVM linker (`wasm-ld`) runs, and so, if we can avoid legalization, we may be able to avoid making any changes to the Wasm file at all. You can get that speedup if you build with `-s WASM_BIGINT`, and do not provide any other flags that require changes to be made. For example, `-O0 -s WASM_BIGINT`, or `-O1 -s WASM_BIGINT`, will work (but at higher optimization levels [the Binaryen optimizer will run](https://emscripten.org/docs/optimizing/Optimizing-Code.html#link-times)).
```

Reply:
Good question, I'm not sure... the original feels better to me, but that may be wrong... maybe someone else will have an opinion.

Reply:
General rule I follow (which AFAIK applies to English just like to other languages) is that commas need to be used around parts of the sentence that are not essential to its meaning (that is, if they're removed, the sentence remains valid). In this regard, e.g. `when using the BigInt integration` or `if we can avoid legalization` give extra context for clarity but otherwise aren't essential to the meaning of the sentence, and it would still be understandable without them.

Reply:
That's good advice, I agree. It does make the sentence have a great deal of commas though, which I've been told is not great either. I did a full rewrite of the paragraph, trying to avoid both, let me know what you think.